### ISSUE TITLE
Per-instance registry?

### ISSUE THREAD

**Issue by anntzer** [seequent/methoddispatch#12]
Per-subclass registries are nice, but an even nicer addition would be per-instance registries, e.g.
```
class C(SingleDispatch):
    @singledispatch
    def foo(self, bar): ...

c = C()
register(c.foo, some_type, some_function)
```

As an example use case, in mplcursors (https://mplcursors.readthedocs.io/en/stable/), I use functools.singledispatch to generate "cursor" implementation for various Matplotlib artist classes, but an end user cannot override the way a specific class is picked without modifying the global singledispatch object.  Having per-instance registries would, well, allow such a per-instance override.

**Comment by tim-mitchell** [comment]
Cool idea.
have subverted the add_overload_method branch for this issue.

**Comment by tim-mitchell** [comment]
implemented in 3.0
